mm2src/coins/utxo/utxo_common.rs

```diff
@@ -2087,18 +2157,72 @@ where
+/// It's used also by watcher to verify that all the taker payment inputs are signed/owned by the taker's pubkey.
+/// The `expected_pub` should be the taker's pubkey in compressed (33-byte) format.
+pub async fn check_all_utxo_inputs_signed_by_pub<T: UtxoCommonOps>(
+    coin: &T,
     tx: &UtxoTx,
     expected_pub: &[u8],
 ) -> Result<bool, MmError<ValidatePaymentError>> {
-    for input in &tx.inputs {
+    let expected_pub =
+        H264::from_slice(expected_pub).map_to_mm(|e| ValidatePaymentError::TxDeserializationError(e.to_string()))?;
+    let mut unsigned_tx: TransactionInputSigner = tx.clone().into();
+
+    for (idx, input) in tx.inputs.iter().enumerate() {
+        let script = Script::from(input.script_sig.clone());
         let pubkey = if input.has_witness() {
+            // Extract the pubkey from a P2WPKH scriptSig.
             pubkey_from_witness_script(&input.script_witness).map_to_mm(ValidatePaymentError::TxDeserializationError)?
+        } else if does_script_spend_p2pk(&script) {
+            // If the transaction is overwintered, we need to set the consensus branch id and the input's amount.
+            // This is needed for the sighash calculation.
+            if unsigned_tx.overwintered {
+                let prev_output_tx_hash = input.previous_output.hash.reversed().into();
+                let prev_output_index = input.previous_output.index as usize;
+                let prev_tx_hex = coin
+                    .as_ref()
+                    .rpc_client
+                    .get_transaction_bytes(&prev_output_tx_hash)
+                    .compat()
+                    .await
+                    .map_err(|e| {
+                        ValidatePaymentError::TxDeserializationError(format!("Failed to get prev tx hex: {e}"))
+                    })?;
+                let prev_tx: UtxoTx = deserialize(prev_tx_hex.0.as_slice())?;
+                let prev_output = prev_tx.outputs.get(prev_output_index).ok_or_else(|| {
+                    ValidatePaymentError::TxDeserializationError(format!(
+                        "Prev tx output index {} out of bounds for tx {}",
+                        input.previous_output.index,
+                        prev_tx.hash()
+                    ))
+                })?;
+                unsigned_tx.inputs[idx].amount = prev_output.value;
+                unsigned_tx.consensus_branch_id = coin.as_ref().conf.consensus_branch_id;
+            }
+            // Verfiy that the P2PK input's scriptSig corresponds to the expected public key.
+            let successful_verification = verify_p2pk_input_pubkey(
+                &script,
+                &Public::Compressed(expected_pub),
+                &unsigned_tx,
+                idx,
+                coin.as_ref().conf.signature_version,
+                coin.as_ref().conf.fork_id,
+            )
+            .map_to_mm(ValidatePaymentError::TxDeserializationError)?;
```

Comment:
It doesn't seem right map validation errors to `TxDeserializationError`.

Reply:
if the sig verification fails, the func returns Ok(false) and not Err().

it returns Err() for similar reasons as `pubkey_from_script_sig` and `pubkey_from_witness_script` i.e. pubkey parsing, scriptpub parsing, signature parsing, etc...